src/StreamJsonRpc/Reflection/RpcTargetInfo.cs

```diff
@@ -0,0 +1,685 @@
+        private readonly JsonRpc jsonRpc;
+
+        /// <summary>
+        /// A collection of target objects and their map of clr method to <see cref="JsonRpcMethodAttribute"/> values.
+        /// </summary>
+        private readonly Dictionary<string, List<MethodSignatureAndTarget>> targetRequestMethodToClrMethodMap = new Dictionary<string, List<MethodSignatureAndTarget>>(StringComparer.Ordinal);
+
+        /// <summary>
+        /// A list of event handlers we've registered on target objects that define events. May be <c>null</c> if there are no handlers.
+        /// </summary>
+        private List<EventReceiver>? eventReceivers;
+
+        /// <summary>
+        /// A lazily-initialized list of objects to dispose of when the JSON-RPC connection drops.
+        /// </summary>
+        private List<object>? localTargetObjectsToDispose;
+
+        internal RpcTargetInfo(JsonRpc jsonRpc)
+        {
+            this.jsonRpc = jsonRpc;
+        }
+
+        private TraceSource TraceSource => this.jsonRpc.TraceSource;
+
+        private object SyncObject => this.targetRequestMethodToClrMethodMap;
+
+        public async ValueTask DisposeAsync()
+        {
+            if (this.localTargetObjectsToDispose is object)
+            {
+                List<Exception>? exceptions = null;
+                foreach (object target in this.localTargetObjectsToDispose)
+                {
+                    // We're calling Dispose on the target objects, so switch to the user-supplied SyncContext for those target objects.
+                    await this.jsonRpc.SynchronizationContextOrDefault;
+
+                    try
+                    {
+                        // Arrange to dispose of the target when the connection is closed.
+                        if (target is System.IAsyncDisposable asyncDisposableTarget)
+                        {
+                            await asyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is Microsoft.VisualStudio.Threading.IAsyncDisposable vsAsyncDisposableTarget)
+                        {
+                            await vsAsyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is IDisposable disposableTarget)
+                        {
+                            disposableTarget.Dispose();
+                        }
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
+                    catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                    {
+                        if (exceptions is null)
+                        {
+                            exceptions = new List<Exception>();
```

Comment:
>exceptions [](start = 28, length = 10)

Can be simplified to `exceptions ??= new List<Exception>();`

Reply:
And it works in C# 8. Nice. :)

Reply:
Same in https://github.com/microsoft/vs-streamjsonrpc/pull/598/files#diff-917aff0ba4f129621f1035d2159950ffe99d4ad6f160676ca183a72efb554122R784-R789